framework/src/modules/dpos_v2/commands/pom.ts

```diff
@@ -34,46 +34,91 @@ import {
 	TokenID,
 	ValidatorsMethod,
 } from '../types';
-import { getPunishmentPeriod } from '../utils';
+import { getDelegateWeight, getPunishmentPeriod } from '../utils';
 import { ValidationError } from '../../../errors';
 import { areDistinctHeadersContradicting } from '../../../engine/bft/utils';
 import { DelegateStore } from '../stores/delegate';
+import { DelegatePunishedEvent } from '../events/delegate_punished';
+import { DelegateBannedEvent } from '../events/delegate_banned';
+import { EligibleDelegatesStore } from '../stores/eligible_delegates';
 
-export class ReportDelegateMisbehaviorCommand extends BaseCommand {
+export class ReportMisbehaviorCommand extends BaseCommand {
 	public schema = pomCommandParamsSchema;
 	private _tokenMethod!: TokenMethod;
 	private _validatorsMethod!: ValidatorsMethod;
 	private _governanceTokenID!: TokenID;
+	private _factorSelfVotes!: number;
 
 	public addDependencies(args: PomCommandDependencies) {
 		this._tokenMethod = args.tokenMethod;
 		this._validatorsMethod = args.validatorsMethod;
 	}
 
-	public init(args: { governanceTokenID: TokenID }) {
+	public init(args: { governanceTokenID: TokenID; factorSelfVotes: number }) {
 		this._governanceTokenID = args.governanceTokenID;
+		this._factorSelfVotes = args.factorSelfVotes;
 	}
 
 	// eslint-disable-next-line @typescript-eslint/require-await
 	public async verify(
 		context: CommandVerifyContext<PomTransactionParams>,
 	): Promise<VerificationResult> {
+		const { chainID, getMethodContext, params, header } = context;
+
 		try {
-			validator.validate(this.schema, context.params);
+			validator.validate(this.schema, params);
 		} catch (err) {
 			return {
 				status: VerifyStatus.FAIL,
 				error: err as Error,
 			};
 		}
 
-		const header1 = BlockHeader.fromBytes(context.params.header1);
-		header1.validate();
+		const currentHeight = header.height;
+		const header1 = BlockHeader.fromBytes(params.header1);
+		const header2 = BlockHeader.fromBytes(params.header2);
+
+		if (!header1.generatorAddress.equals(header2.generatorAddress)) {
+			throw new Error('Different generator address never contradict to each other.');
+		}
+
+		const delegateAddress = header1.generatorAddress;
+		const delegateSubStore = this.stores.get(DelegateStore);
+		const delegateAccount = await delegateSubStore.get(context, delegateAddress);
+
+		const { generatorKey } = await this._validatorsMethod.getValidatorKeys(
+			getMethodContext(),
+			header1.generatorAddress,
+		);
+
+		header1.validateSignature(generatorKey, chainID);
+		header2.validateSignature(generatorKey, chainID);
+
+		const maxPunishableHeight = Math.max(
+			Math.abs(header1.height - currentHeight),
+			Math.abs(header2.height - currentHeight),
+		);
+
+		if (maxPunishableHeight >= LOCKING_PERIOD_SELF_VOTES) {
+			throw new Error('Locking period has expired.');
```

Comment:
From what I've seen in other commands, the verification error is always [?] returned in this format:

```typescript
return {
    status: VerifyStatus.FAIL,
    error: new Error('Something something error error'),
};
```

On multiple occasions in this file, the verification errors are implemented with just a simple throw. What was the reason for that?

Reply:
Probably this is my mistake, in general it's better to throw if there is not special status, but we mixed up.
I will create another issue for this